File: Strategy.py
"""
Strategy
Define a family of algorithms, encapsulate each one, and make them interchangeable.
"""

from abc import ABC, abstractmethod


class Strategy(ABC):
    @abstractmethod
    def method(self):
        pass

class Strategy1(Strategy):
    def method(self):
        print("Strategy 1")

class Strategy2(Strategy):
    def method(self):
        print("Strategy 2")

class Context:
    def __init__(self):
        self.strategy: None|Strategy = None

    def setStrategy(self, strategy):
        self.strategy = strategy

    def method(self):
        if self.strategy:
            self.strategy.method()


if __name__ == "__main__":
    c = Context()

    c.setStrategy(Strategy1())
    c.method()
    
    c.setStrategy(Strategy2())
    c.method()

File: Strategy.js
/*****
 * Strategy
 * Define a family of algorithms, encapsulate each one, and make them interchangeable.
*****/


class Strategy { method() {} }

class Strategy1 extends Strategy { method() { console.log("Strategy 1"); } }

class Strategy2 extends Strategy { method() { console.log("Strategy 2"); } }

class Context {
    constructor() { this.strategy = null; }

    setStrategy(strategy) { this.strategy = strategy; }
    method() { if (this.strategy) this.strategy.method(); }
}


const c = new Context();

c.setStrategy(new Strategy1());
c.method();

c.setStrategy(new Strategy2());
c.method();
